File: Keeping/Keeping/KCProgressButton.m
//
//  KCProgressButton.m
//  Keeping
//
//  Created by 宋 奎熹 on 2017/10/17.
//  Copyright © 2017年 宋 奎熹. All rights reserved.
//

#import "KCProgressButton.h"
#import "KCEdgeInsetLabel.h"
#import "Utilities.h"

@implementation KCProgressButton

- (void)setTitle:(NSString *)title forState:(UIControlState)state {
    [super setTitle:title forState:state];
    NSArray *countArr = [title componentsSeparatedByString:@" / "];
    [self addMaskLabelWithProgress:[(countArr[0]) floatValue] / [(countArr[1]) floatValue]];
    [self setNeedsLayout];
    [self setNeedsDisplay];
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {

    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder{
    return [super initWithCoder:aDecoder];
}

- (void)setProgressWithFinished:(int)finished andTotal:(int)total {
    float progress = 0.0;
    if (total > 0) {
        progress = (float)finished / (float)total;
    }
    [self setTitle:[NSString stringWithFormat:@"%d / %d", finished, total] forState:UIControlStateNormal];
    NSLog(@"title %@", self.titleLabel.text);
    [self addMaskLabelWithProgress:progress];
}

- (void)addMaskLabelWithProgress:(float)progress {
    NSLog(@"progress %f", progress);
    
    if (self.maskLabel != nil) {
        [self.maskLabel removeFromSuperview];
    }
    
    self.maskLabel = [[KCEdgeInsetLabel alloc] initWithFrame:CGRectMake(0,
                                                                        0,
                                                                        self.frame.size.width * progress,
                                                                        self.frame.size.height)
                                                andEdgeInset:UIEdgeInsetsMake(0, self.titleLabel.frame.origin.x, 0, 0)];
    self.maskLabel.text = self.titleLabel.text;
    self.maskLabel.font = self.titleLabel.font;
    self.maskLabel.backgroundColor = [Utilities getColor];
    self.maskLabel.textColor = [UIColor whiteColor];
    self.maskLabel.lineBreakMode = NSLineBreakByCharWrapping;
    self.maskLabel.textAlignment = self.titleLabel.textAlignment;
    self.maskLabel.layer.cornerRadius = self.layer.cornerRadius;
    self.maskLabel.layer.masksToBounds = YES;
    [self addSubview:self.maskLabel];
}

@end

File: Keeping/Keeping/KCProgressButton.swift
//
//  KCProgressButton.swift
//  KCProgressButton
//
//  Created by 宋 奎熹 on 2017/9/9.
//  Copyright © 2017年 宋 奎熹. All rights reserved.
//

import UIKit

class KCProgressButton: UIButton {
    
    /// 下载完成时的 Title
    open var finishDownloadTitle: String = "全部完成"
    
    /// 边框宽度
    open var borderWidth: Float = 0.0 {
        didSet {
            layer.borderWidth = CGFloat(borderWidth)
            layer.borderColor = titleColor(for: .normal)?.cgColor
        }
    }
    
    /// 文字的格式函数
    open var valueFormatter: (_: Float) -> String = {
        return String(format: "%.2f", $0 * 100)
    }
    
    /// 内部作 Mask 的 Label
    private var maskLabel: KCEdgeInsetLabel?
    
    //MARK: - Initializers
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        addMaskLabel(with: titleLabel?.text ?? "", progress: 0)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    //MARK: - Set Progress
    
    func setProgress(finished: Int, total: Int) {
        var progress: Float = 0.0
        if total > 0 {
            progress = Float(finished) / Float(total)
        }
        addMaskLabel(with: "\(finished) / \(total)", progress: progress)
    }
    
    private func addMaskLabel(with title: String, progress: Float) {
        print("Add Mask Label")
        if maskLabel != nil {
            maskLabel?.removeFromSuperview()
        }
        
        setTitle(title, for: .normal)
        let titleLabelFrame = titleLabel?.frame
        maskLabel = KCEdgeInsetLabel(frame: CGRect(x: 0,
                                                   y: 0,
                                                   width: self.frame.width * CGFloat(progress),
                                                   height: self.frame.height),
                                     edgetInset: UIEdgeInsetsMake(0,
                                                                  (titleLabelFrame?.origin.x)!,
                                                                  0,
                                                                  0))
        maskLabel?.text = title
        maskLabel?.font = titleLabel?.font
        maskLabel?.backgroundColor = Utilities.getColor()
        maskLabel?.textColor = .white
        maskLabel?.lineBreakMode = .byCharWrapping
        maskLabel?.textAlignment = (titleLabel?.textAlignment)!
        maskLabel?.layer.cornerRadius = layer.cornerRadius
        maskLabel?.layer.masksToBounds = true
        addSubview(maskLabel!)
    }
    
    /// 内部作 Mask 的 Label
    class KCEdgeInsetLabel: UILabel {
        
        /// 内部 Label 的内边距
        private var textEdgeInsets: UIEdgeInsets = .zero
        
        override init(frame: CGRect) {
            super.init(frame: frame)
        }
        
        convenience init(frame: CGRect, edgetInset: UIEdgeInsets) {
            self.init(frame: frame)
            textEdgeInsets = edgetInset
        }
        
        required init?(coder aDecoder: NSCoder) {
            fatalError("init(coder:) has not been implemented")
        }
        
        /// 重载此方法以达到内边距与 Button 的 TitleLabel 内边距相同的目的
        override func drawText(in rect: CGRect) {
            super.drawText(in: UIEdgeInsetsInsetRect(rect, textEdgeInsets))
        }
        
    }
    
}
